---
title: "Screening of contaminant features"
author: Yanxian Li
date: "`r Sys.time()`"
output: 
  html_document: 
    self_contained: true
    toc: true
    toc_float: true
    toc_depth: 2
  pdf_document: 
    latex_engine: xelatex
  word_document: default
---

```{r style, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
options(width=150) 
opts_chunk$set(message=FALSE, error=FALSE, warning=FALSE)
```

# Load packages
```{r load-package}
library(here) # [CRAN] A replacement for 'file.path()', locating the files relative to the project root.
library(tidyverse) # [CRAN] Easily Install and Load the 'Tidyverse' 
library(cowplot) # [CRAN] # Streamlined Plot Theme and Plot Annotations for 'ggplot2' 
library(ggstatsplot) # [CRAN] # 'ggplot2' Based Plots with Statistical Details  
```

# Introduction
The screening will be based on two typical characteristics of contaminant sequences as outlined in the paper by [Davis et al. (2018)](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0605-2): they are likely to have frequencies that inversely correlate with sample DNA concentration and are likely to have higher prevalence in control samples than in true samples. The authors developed an R package, [*decontam*](https://github.com/benjjneb/decontam), for removing contaminant sequences in the marker-gene and metagenomics data. The package, however, does not make use of positive controls for the identification of contaminant sequences. As removing of features may heavily affect downstream analyses, we'll do it by manual screening based on the aforementioned principles.

# Import and tidy data 
## Metadata 
```{r import_metadata}
metadata <- read_tsv(here("data", "metadata.tsv"), comment = "#q2:type") %>%
  rename(SampleID = `#SampleID`) 
```

## Feature table 
```{r import_tidy_table}
tab <- read_tsv(here("data", "qiime2", "decontam", "table-no-chlo-mito-lowPre-with-phyla-tax-rel.tsv"), 
                skip = 1)
tab <- tab %>%
  # sum up features found in the negative and positive controls
  mutate(Ctrl = rowSums(.[grep("EB|LB|M", names(.))], na.rm = TRUE)) %>% 
  # remove features not present in the negative or positive controls
  filter(Ctrl > 0) %>% 
  # sort features by total relative abundance in the negative and positive controls
  arrange(desc(Ctrl)) %>% 
  # remove the newly created variable "Ctrl"
  select(-Ctrl) %>% 
  # the following 4 lines of codes transpose the dataframe
  rownames_to_column %>% 
  mutate(rowname = factor(rowname, levels = unique(rowname))) %>% 
  gather(var, value, -rowname) %>%
  spread(rowname, value) %>%
  # use first row as the column name
  `colnames<-`(.[1, ]) %>% 
  # remove the first row
  slice(-1) %>% 
  rename(SampleID = `#OTU ID`)
```

Extract the taxonomy and convert to long format
```{r extract_taxonomy}
tax <- tab %>%
  filter(SampleID == "taxonomy") %>%
  t() %>%
  data.frame() %>%
  # use first row as the column name
  `colnames<-`(.[1, ]) %>% 
  slice(-1) %>% 
  # make 80 (number of samples) replicates of each row 
  uncount(nrow(tab)-1) 
```

## Merge feature table and metadata 
```{r merge_table}
tab_merged <- inner_join(metadata, tab, by = "SampleID") %>%
  arrange(match(SampleType, c("REF-DIC", "REF-DIM", "IM-DIC", "IM-DIM", 
                              "Blank-library", "Blank-extraction", "Mock"))) %>%
  gather(featureID, abundance, 74:ncol(.)) %>%
  # convert "abundance" to a numeric variable as percentage
  mutate(abundance = as.numeric(abundance), abundance = abundance * 100) %>% 
  # add taxonomy after the featureID column 
  add_column(tax = tax[, 1], .after = "featureID") %>% 
  as.data.frame()
```

Convert featureID, SampleID and SampleType to factors
```{r as_factor}
tab_merged$featureID <- factor(tab_merged$featureID, unique(tab_merged$featureID)) 

tab_merged$SampleID <- factor(tab_merged$SampleID, unique(tab_merged$SampleID)) 

tab_merged$SampleType <- factor(tab_merged$SampleType, c("REF-DIC", "REF-DIM", "IM-DIC", "IM-DIM",
                                                         "Blank-library", "Blank-extraction", "Mock")) 
```

# Feature prevalence
We'll use barplot to visualize the abundance and prevalence of the features found in the positive and negative controls. 
```{r make_barplot, results='hide'}
# Split the dataframe by "featureID"
tab_merged_spl <- split(tab_merged, f = tab_merged$featureID)

# Make barplot for each feature
pdf(here("data", "qiime2", "decontam", "taxa_barplot_unstacked.pdf"), width = 16, height = 10) 

lapply(
  seq_along(tab_merged_spl), # add index to the elements in the list 
  function(x) 
  {
    # Make bar plots including samples and negative controls
    p1 <- tab_merged_spl[[x]] %>%
      filter(SampleType != "Mock") %>%
      ggplot(aes(x = SampleID, y = abundance, fill = SampleType)) +
        geom_bar(stat = "identity", position = position_dodge()) +
        scale_y_continuous(limits = c(0, NA), expand = expand_scale(mult = c(0, 0.1))) + 
        labs(title = unique(tab_merged_spl[[x]][, "tax"]), y = "relative abundance (%)") +
        scale_fill_brewer(palette = "Dark2", drop = F) +
        guides(fill = guide_legend(nrow = 1)) +
        theme_bw() +
        theme(plot.title = element_text(size = 12),
              axis.title.x = element_text(size = 12),
              axis.text.x = element_text(angle = 45, hjust = 1),
              axis.title.y = element_text(size = 12),
              legend.position = "bottom")
    
    # Make bar plots including mock only
    p2 <- tab_merged_spl[[x]] %>%
      filter(SampleType == "Mock") %>%
      ggplot(aes(x = SampleID, y = abundance, fill = SampleType)) +
      geom_bar(stat = "identity", position = position_dodge()) +
      scale_y_continuous(limits = c(0, NA), expand = expand_scale(mult = c(0, 0.1))) + 
      labs(x = "", y = "") +
      scale_fill_brewer(palette = "Dark2", drop = F) +
      guides(fill = guide_legend(nrow = 1)) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "bottom")
    
    # Assemble plots
    plot_grid(p1, p2 + theme(legend.position = "none"), nrow = 1, align = 'h', 
              axis = "bt", rel_widths = c(15, 1))
  }
)

dev.off() 
```

# Correlation analysis
We'll use Cq values from the qPCR assays as proxies of microbial DNA concentration in the samples. A clear positive correlation between the feature abundances and Cq values is indicative of contamination.

## Cq values distribution
Before we proceed with correlation analysis, we'll Check Cq values and see if they make sense. Normally, we expect lower Cq values for gut contents and mock, and higher Cq values for mucosa and negative controls.
```{r plot_Cq}
set.seed(1910) # to get same jitters every time
tab_merged %>% 
  # Get Cq values from one of the features
  filter(featureID == unique(.$featureID)[1]) %>%
  drop_na(qPCRCqValue) %>%
  ggplot(aes(x = SampleType, y = qPCRCqValue)) +
  geom_jitter(aes(colour = SampleType), shape = 16, position = position_jitter(0.1)) +
  # add mean plus/minus 1SD 
  stat_summary(fun.data = "mean_sdl", fun.args = list(mult = 1), geom = "pointrange") + 
  labs(x = "", y = "Cq value", title = "Quantification of 16S rRNA gene by qPCR") +
  theme_cowplot() +
  scale_fill_brewer(palette = "Dark2") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(here("exploratory", "qPCR_Cq_values.pdf"), width = 6, height = 6, units = "in", dpi = 300)
```

Everything looks normal. We proceed to correlation analysis.

## Correlation analysis 
```{r plot_correlation, results='hide'}
# Exclude samples with zero count of features
tab_merged <- filter(tab_merged, abundance != 0)

# Split the dataframe by featureID
tab_merged_spl <- split(tab_merged, f = tab_merged$featureID)

# Make correlation plots
pdf(here("data", "qiime2", "decontam", "correlation_analysis.pdf"), width = 16, height = 12) 

lapply(
  seq_along(tab_merged_spl), # add index to elements in the list 
  function(x) 
  {
    # Extract taxonomy and featureID as title and subtitle, respectively
    main <- ggdraw() + draw_label(unique(tab_merged_spl[[x]]$tax), fontface='bold')
    
    # Correlation using all samples
    p_all <- tab_merged_spl[[x]]%>% 
      ggplot(aes(x = qPCRCqValue, y = abundance)) +
      geom_point(aes(color = SampleType)) +
      geom_smooth(method = "lm") +
      scale_y_continuous(limits = c(0, NA), expand = expand_scale(mult = c(0, 0.1))) + 
      labs(x = "Cq value", y = "relative abundance (%)", title = "Correlation with all samples") +
      scale_fill_brewer(palette = "Dark2") +
      guides(colour = guide_legend(nrow = 1)) +
      theme_bw() +
      theme(plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
            axis.title.x = element_text(size = 12),
            axis.title.y = element_text(size = 12),
            legend.position = "bottom")
    
    # Corraltion analysis within each sample type
    p_sub <- tab_merged_spl[[x]] %>%
      filter(!SampleType %in% c("Blank-library", "Blank-extraction", "Mock")) %>%
      # make a list column grouped by 'SampleType'
      group_split(SampleType, keep = T) %>% 
      map(function(x)
        # turn off stats if the number of observation is less than 6
        if(sum(!is.na(x$abundance)) < 6){ 
          ggscatterstats(data = x,
                         x = qPCRCqValue, 
                         y = abundance,
                         type = "spearman", # type of test that needs to be run
                         conf.level = 0.95, # confidence level
                         xlab = "Cq value", # label for x axis
                         ylab = "relative abundance (%)", # label for y axis
                         title = paste0("Correlation within the sample type: ", unique(x$SampleType)),
                         k = 3, # no. of decimal places in the results
                         marginal = F, # whether show ggExtra::ggMarginal() plots or not
                         results.subtitle = F, # turn off stats
                         messages = F) # turn off messages and notes
        } else {
          # turn on stats if the number of observation is no less than 6
          ggscatterstats(data = x,
                         x = qPCRCqValue, 
                         y = abundance,
                         type = "spearman", 
                         conf.level = 0.95, 
                         xlab = "Cq value", 
                         ylab = "relative abundance (%)", 
                         title = paste0("Correlation within the sample type: ", unique(x$SampleType)),
                         k = 3, 
                         marginal = F, 
                         messages = F) 
        }
      )
    
    # Assemble plots
    panel_1 <- plot_grid(main, p_all, ncol = 1, rel_heights = c(1, 12)) 
    panel_2 <- plot_grid(plotlist = p_sub, ncol = 2, rel_heights = c(1, 1)) 
    plot_grid(panel_1, panel_2, ncol = 1, rel_heights = c(1, 1.2))
  }
)

dev.off() 
```

# List of contaminant sequences 
Based on the evaluation of the taxa bar plots, correlation plots and our prior knowledge of common contaminant taxa found in our lab, the following features are considered as contaminant sequences in the samples.
```{r contam_sample}
contam_sample <- tab_merged %>%
  select(featureID, tax) %>%
  distinct() %>%
  filter(row_number() %in% c(1,2,4,5,6,8,10,11,14,16,18,21,23,26,29,30,32,33,34,35,36,37,38,39,40,42,
                             43,45,46,47,48,50,51,53,54,55,56,58,59,60,62,63,64,66,71,76,88,90,104))
```

The mock, [ZymoBIOMICS D6300](https://www.zymoresearch.com/collections/zymobiomics-microbial-community-standards/products/zymobiomics-microbial-community-standard), comes with a guaranteed impurity level of < 0.01% (by DNA abundance). Therefore, as long as we observe any alien taxa present at > 0.01% in the mock, we can conclude that they are introduced during by the workflow. The following features are considered as contaminant features in the mock.
```{r contam_mock}
contam_mock <- tab_merged %>%
  select(featureID, tax) %>%
  distinct() %>%
  filter(row_number() %in% c(1,6,21,23,26,30,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,
                             52,53,54,55,56,58,59,60,61,62,63,64,65,66,67,69,71,73,74,76,85,88,90,91,92))                
```

# Export contaminant sequence IDs
Now we export the featureIDs of contaminant sequences for feature table filtering.
```{r export_contam}
# Contaminant features for samples
write.table(contam_sample, 
            file = here("data", "qiime2", "decontam", "contaminant-features-sample.tsv"), 
            quote = FALSE, 
            sep = '\t',
            row.names = FALSE)

# Contaminant features for the mock
write.table(contam_mock, 
            file = here("data", "qiime2", "decontam", "contaminant-features-mock.tsv"), 
            quote = FALSE, 
            sep = '\t',
            row.names = FALSE)
```

# Session information
```{r session-info}
sessionInfo()
```
