---
title: "Taxonomic analysis"
author: Yanxian Li
date: "`r Sys.time()`"
output: 
  html_document: 
    self_contained: true
    toc: true
    toc_float: true
    toc_depth: 2
  pdf_document: 
    latex_engine: xelatex
  word_document: default
---
  
<style>
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>

```{r style, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
options(width=150) 
opts_chunk$set(message=FALSE, error=FALSE, warning=FALSE)
```

# Getting ready
Load packages
```{r load-package}
library(here) # [CRAN] A replacement for 'file.path()', locating the files relative to the project root
library(tidyverse) # [CRAN] Easily install and load the 'Tidyverse'   
library(cowplot)  # [CRAN] Streamlined plot theme and plot annotations for 'ggplot2'
library(PerformanceAnalytics) # [CRAN] Collection of econometric functions for performance and risk analysis
library(RColorBrewer) # [CRAN] ColorBrewer Palettes
library(qiime2R) # [github::jbisanz/qiime2R] Import qiime2 artifacts to R 
library(MicrobeR) # [github::jbisanz/MicrobeR] Handy functions for downstream visualization fo microbiome data
library(microbiome) # [Bioconductor] Microbiome analytics
library(phyloseq) # [Bioconductor] Handling and analysis of high-throughput microbiome census data 
library(DT) # [CRAN] An R interface to the DataTables library
library(gt) # [github::rstudio/gt] # Easily Create Presentation-Ready Display Tables
library(venn) # [CRAN] Make Venn's diagram
```

Load functions
```{r}
source(here("code", "functions", "make_taxa_barplot.R"))
```

Import data
```{r load-data}
load(here("data", "qiime2R", "phyloseq.RData"))

# Extract feature table, taxonomy and metadata from the phyloseq object
count_tab <- as.data.frame(otu_table(phyloseq)) 
tax_tab <- tax_table(phyloseq) %>% as("matrix") %>% as.data.frame()
metadata <- data.frame(sample_data(phyloseq), check.names = FALSE)
```

# Overview of taxonomy assignments 
First of all, let's look at the coverage of taxonomy assignments at different levels.
```{r}
tax_tab %>%
  gather("Rank", "Name", rank_names(phyloseq)) %>%
  group_by(Rank) %>%
  # Empty taxonomic ranks may be na or strings containing "uncultured" or "Ambiguous_taxa"
  summarize(ASVs_classified = sum(!is.na(Name) & !grepl("uncultured|Ambiguous|metagenome", Name))) %>%
  mutate(Frac_classified = ASVs_classified / ntaxa(phyloseq),
         Frac_classified = ifelse(Frac_classified == 1, "100", round(Frac_classified * 100, 1)),
         Frac_classified = paste(Frac_classified, "%"),
         Rank = factor(Rank, rank_names(phyloseq))) %>%
  arrange(Rank) %>%
  datatable(options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:3)))))
```

We can tell that the majority of ASVs were assigned at the genus level whereas only 23% of ASVs got a species-level annotation.

# Taxonomic analysis: control samples
## Mock 
Import and tidy expected mock composition 
```{r}
mock_exp <- read_tsv(here("data", "reference", "mock_expected.tsv"))   
mock_exp <- select(mock_exp, -"AqFl2-M2") %>%
  rename(tax_exp = "#OTU ID", MOCK = "AqFl2-M1") %>%
  arrange(tax_exp) %>%
  mutate(tax_exp = gsub("D_0.*D_6__", "", tax_exp)) # prune taxonomy to the lowest level
```

Import and tidy observed mock composition 
```{r}
mock_obs <- read_qza(here("data", "qiime2", "quality-control", "mock-observed-l7-rel.qza"))
mock_obs <- mock_obs$data %>%
  as.data.frame() %>%
  rownames_to_column("tax_obs") %>%
  rename(MOCK1 = "AqFl2-M1", MOCK2 = "AqFl2-M2") %>%
  # prune taxonomy to the lowest level
  mutate(tax_obs = gsub("D_0.*D_6__", "", tax_obs),  
         tax_obs = gsub("D_0.*D_5__|;__", "", tax_obs))
```

Merge expected and observed mock composition 
```{r}
mock_merged <- bind_cols(mock_exp, mock_obs) %>%
  mutate(Gram_staining = c("(G+)", "(G+)", "(G+)", "(G+)", "(G+)", "(G-)", "(G-)", "(G-)"),
         "taxa" = paste(tax_exp, Gram_staining, "/", tax_obs)) %>%
  select(taxa, MOCK, MOCK1, MOCK2) %>%
  mutate(taxa = factor(.$taxa, levels = unique(.$taxa))) %>% 
  # use percentage as unit for relative abundance
  mutate_if(is.numeric, ~(100 * .x)) %>% 
  mutate_if(is.numeric, ~round(.x, 1))

mock_merged %>%
  rename("Taxa (expected / observed)" = taxa) %>%
  datatable(options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:4)))))
```

Make correlation plot
```{r}
chart.Correlation(mock_merged[,2:4], histogram = FALSE, pch = 19)
```

Make stacked barplot
```{r}
# use italic font for genus/species names in the legend
labs_mock <- c(
  expression(paste(italic("Bacillus subtilis"), "(G+) / ", italic("Bacillus"))), 
  expression(paste(italic("Listeria monocytogenes"), "(G+) / ", italic("Listeria monocytogenes"))),
  expression(paste(italic("Staphylococcus aureus"), "(G+) / ", italic("Staphylococcus aureus"))),
  expression(paste(italic("Enterococcus faecalis"), "(G+) / ", italic("Enterococcus faecalis"))),
  expression(paste(italic("Lactobacillus fermentum"), "(G+) / ", italic("Lactobacillus fermentum"))),
  expression(paste(italic("Escherichia coli"), "(G-) / ", italic("Escherichia-Shigella"))),
  expression(paste(italic("Salmonella enterica"), "(G-) / ", italic("Salmonella"))),
  expression(paste(italic("Pseudomonas aeruginosa"), "(G-) / ", italic("Pseudomonas"))))

# Make barplot
p_mock <- mock_merged %>% 
  gather(key = "type", value = "abundance", -taxa) %>%
  ggplot(aes(type, abundance)) + 
  geom_bar(aes(fill = taxa), stat = "identity", width = 0.5) + 
  annotate("segment", x = 1.75, xend = 3.25, y = 102, yend = 102) +
  annotate("text", x = c(1, 2.5), y = c(105, 105), label = c("Expected", "Observed"), size = 4) +
  scale_y_continuous(limits = c(0, 105), breaks = 0:10*10, expand = expansion(mult = c(0, 0.05))) +
  labs(x = "", y = "Relative abundance (%)", fill = "Taxa (expected (Gram stain) / observed)") +
  scale_fill_manual(values = brewer.pal(8, "Paired"), labels = labs_mock) +
  theme_minimal() +
  theme(panel.grid.major.y = element_line(size = 0.5, linetype = 'dashed', colour = "black"),
        legend.justification = "top", # move legend to top right position 
        legend.text.align = 0) # align legend text to left
p_mock
```

## Negative controls 
Import and tidy data for making taxa barplot.
```{r}
# Import contaminating sequences found in the control samples
neg <- read_csv(here("result", "tables", "Table S1.csv"))

# tidy table
neg <- select(neg, -contains("Mock")) %>%
  # the next 3 lines remove features not present in the negative controls
  mutate(taxa_sum = rowSums(.[grepl("blank", names(.))], na.rm = TRUE)) %>% 
  filter(taxa_sum > 0) %>% 
  select(-taxa_sum) %>% 
  # the next 9 lines produce the lowest level of taxonomic rank for each feature
  separate(Taxonomy, sep = ";", c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>% 
  mutate(Class = ifelse(grepl("NA|uncultured|Ambiguous|metagenome", Class), Phylum, Class), 
         Order = ifelse(grepl("NA|uncultured|Ambiguous|metagenome", Order), Class, Order),
         Family = ifelse(grepl("NA|uncultured|Ambiguous|metagenome", Family), Order, Family),
         Genus = ifelse(grepl("NA|uncultured|Ambiguous|metagenome", Genus), Family, Genus),
         Species = ifelse(grepl("NA|uncultured|Ambiguous|metagenome", Species), Genus, Species)) %>%
  select(-(Kingdom:Genus)) %>%
  rename(taxonomy = Species) %>%
  # remove taxonomic prefix from features with genus/species level annotation
  mutate(taxonomy = gsub("g__|s__", "", taxonomy),
         taxonomy = factor(taxonomy, levels = unique(taxonomy))) %>%
  # the next 2 lines agglomerate features by taxonomic labels
  group_by(taxonomy) %>%
  summarise_if(is.numeric, ~sum(.x)) %>%
  # use percentage as unit for relative abundance
  mutate_if(is.numeric, ~.x * 100)
```

Display taxa in the negative controls
```{r}
neg %>%
  mutate_if(is.numeric, ~ifelse(.x != 0, round(.x, 4), .x)) %>%
  datatable(options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:7))))) 
```

Make stacked barplot
```{r}
p_neg <- neg %>%
  pivot_longer(-taxonomy, names_to = "SampleName", values_to = "Abundance") %>%
  mutate(SampleName = gsub("Extraction-blank", "EB", SampleName), 
         SampleName = gsub("Library-blank", "LB", SampleName)) %>%
  ggplot(aes(SampleName, Abundance)) + 
  geom_bar(aes(fill = taxonomy), stat = "identity", width = 0.5) + 
  annotate("segment", x = c(0.75, 4.75), xend = c(4.25, 6.25), y = c(102, 102), yend = c(102,102)) +
  annotate("text", x = c(2.5, 5.5), y = c(105, 105), size = 4, label = c("Extraction blank", "Library blank")) +
  scale_y_continuous(limits = c(0, 105), breaks = 0:10*10, expand = expansion(mult = c(0, 0.05))) +
  labs(x = "", y = "Relative abundance (%)", fill = "Taxa") +
  scale_fill_manual(values = brewer.pal(12, "Paired"),
                    guide = guide_legend(label.theme = element_text(face = "italic", size = 9))) +
  theme_minimal() +
  theme(panel.grid.major.y = element_line(size = 0.5, linetype = 'dashed', colour = "black"),
        legend.justification = "top") # move legend to top right position (default is on the right)
p_neg
```

## Assemble plots 
```{r, fig.width=6, fig.height=8}
plot_grid(p_mock, p_neg, labels = "AUTO", ncol = 1) 

# Export the assembled plot
ggsave(here("result", "figures", "Figure S2.tiff"), width = 6, height = 8, 
       units = "in", dpi = 300, compression = "lzw")
```

# Taxonomic analysis: biological samples  
## Phylum-level 
Collapse feature table at phylum level
```{r}
tab_phy <- Summarize.Taxa(count_tab, tax_tab)$Phylum %>%
  # the following 3 lines of codes prune the taxonomy to contain phylum names only
  rownames_to_column("tax") %>%
  mutate(tax = gsub("k__.*p__", "", tax)) %>%
  column_to_rownames("tax")
```

Make taxa barplot
```{r}
# Plot taxa on individual basis
p_phy_ind <- make_taxa_barplot(table = tab_phy, 
                               metadata = metadata, 
                               group_by = SampleType, 
                               ntaxa = 10,
                               nrow = 1,
                               plot_mean = FALSE, 
                               cluster_sample = FALSE,
                               sample_label = FishID,
                               italize_taxa_name = TRUE)
# Plot taxa using group means
p_phy_mean <- make_taxa_barplot(table = tab_phy, 
                                metadata = metadata, 
                                group_by = SampleType, 
                                ntaxa = 10,
                                nrow = 1,
                                plot_mean = TRUE, 
                                cluster_sample = FALSE,
                                #sample_label = FishID,
                                italize_taxa_name = TRUE)
                                
p_phy_mean <- p_phy_mean + labs(x = "", y = "")
```

## Genus-level 
Collapse feature table at genus level
```{r}
tab_gen <- Summarize.Taxa(count_tab, tax_tab)$Genus %>%
  rownames_to_column("tax") %>%
  # the following 6 lines of codes produce genus-level taxonomy for each feature
  separate(tax, sep = ";", c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")) %>% 
  mutate(Class = ifelse(Class == "NA"|grepl("uncultured|Ambiguous|metagenome", Class), Phylum, Class),
         Order = ifelse(Order == "NA"|grepl("uncultured|Ambiguous|metagenome", Order), Class, Order),
         Family = ifelse(Family == "NA"|grepl("uncultured|Ambiguous|metagenome", Family), Order, Family),
         Genus = ifelse(Genus == "NA"|grepl("uncultured|Ambiguous|metagenome", Genus), Family, Genus)) %>%
  select(-(Kingdom:Family)) %>%
  # remove the "g__" prefix from genus names
  mutate(Genus = gsub("g__", "", Genus)) %>%
  # the following 4 lines of codes collapse features by taxonomy
  gather("SampleID", "counts", -Genus) %>%
  group_by(SampleID, Genus) %>%
  summarize(counts_colsum = sum(counts)) %>%
  spread("SampleID", "counts_colsum") %>%
  column_to_rownames("Genus")
```

Make taxa barplot showing top 10 genera
```{r}
# Plot taxa on individual basis
p_gen_ind <- make_taxa_barplot(table = tab_gen, 
                               metadata = metadata, 
                               group_by = SampleType, 
                               ntaxa = 10,
                               nrow = 1,
                               plot_mean = FALSE, 
                               cluster_sample = FALSE,
                               sample_label = FishID,
                               italize_taxa_name = TRUE)
# Plot taxa using group means
p_gen_mean <- make_taxa_barplot(table = tab_gen, 
                                metadata = metadata, 
                                group_by = SampleType, 
                                ntaxa = 10,
                                nrow = 1,
                                plot_mean = TRUE, 
                                cluster_sample = TRUE,
                                sample_label = FishID,
                                italize_taxa_name = TRUE)
                                
p_gen_mean <- p_gen_mean + labs(x = "", y = "")
```

## Assemble plots 
```{r, fig.width=14, fig.height=10}
plot_grid(p_phy_ind + theme(legend.position = "none"), 
          p_phy_mean + theme(axis.text.x = element_text(angle = 45, hjust = 1), 
                             plot.margin = margin(l = -0.8, unit = "cm"),
                             legend.box.margin=margin(l = -10, r = 3)),
          p_gen_ind + theme(legend.position = "none"), 
          p_gen_mean + theme(axis.text.x = element_text(angle = 45, hjust = 1),
                             plot.margin = margin(l = -0.8, unit = "cm"), 
                             legend.box.margin=margin(l = -10, r = 3)),
          labels = c("A", "", "B", ""), ncol = 2, align = 'hv', axis = "tb", 
          rel_widths = c(3, 1, 3, 1)) 

# Export the plot
ggsave(here("result", "figures", "Figure 1.tiff"), width = 14, height = 10, 
       units = "in", dpi = 300, compression = "lzw")
```
# Core microbiome 
Calculate feature prevalence
```{r}
# Compute feature prevalence
prev_rd <- subset_samples(phyloseq, SampleType == "REF-DID") %>% prevalence() 
prev_rm <- subset_samples(phyloseq, SampleType == "REF-DIM") %>% prevalence() 
prev_id <- subset_samples(phyloseq, SampleType == "IM-DID") %>% prevalence() 
prev_im <- subset_samples(phyloseq, SampleType == "IM-DIM") %>% prevalence() 
```

Get core features that are present in at least 80% samples under different sample types
```{r}
core_taxa <- cbind.data.frame(prev_rd, prev_rm, prev_id, prev_im) %>%
  rownames_to_column("featureID") %>%
  # get core features based on 80% prevalence threshold
  filter(prev_rd >= 0.8|prev_rm >= 0.8|prev_id >= 0.8|prev_im >= 0.8)
```

Add taxonomy to core features
```{r}
core_taxa_tab <- rownames_to_column(tax_tab, "featureID") %>%
  inner_join(core_taxa, by = "featureID") %>%
  rename("REF-DID" = prev_rd, "REF-DIM" = prev_rm, "IM-DID" = prev_id, "IM-DIM" = prev_im) %>%
  mutate(prev_all = rowSums(.[9:12])) %>%
  arrange(desc(prev_all)) %>%
  mutate_if(is.numeric, ~ifelse(.x == 1, 100, round(.x * 100, 1))) %>%
  mutate_if(is.numeric, ~paste0(.x, "%")) %>%
  mutate_if(is.character, ~gsub("*.__|uncultured.*|Ambiguous.*|metagenome", "", .x)) %>%
  map_df(~replace_na(.x, "")) %>%
  select(-featureID, -Kingdom, -prev_all) 

core_taxa_tab %>%
  datatable(options = list(columnDefs = list(list(className = 'dt-left', targets = c(0:7)))))
```

```{r, include=FALSE}
# Export core taxa table as Table S2
core_taxa_tab %>%
  gt() %>%
  tab_header(title = "Table S2. The prevalence of core ASVs in different sample types.") %>%
  tab_spanner(label = "Taxonomy", columns = 1:6) %>%
  tab_spanner(label = "Prevalence", columns = 7:ncol(core_taxa_tab)) %>%
  cols_align(align = "center") %>%
  tab_source_note("The core ASVs were computed using a prevalence threshold at 80%.") %>%
  tab_source_note("Abbreviations: REF, reference diet; IM, insect meal diet; 
                  DID, distal intestine digesta; DIM, distal intestine mucosa.") %>%
  # the top/bottom border color can't be modified
  tab_style(style = list(cell_text(weight = "bold", size = 14), cell_borders(sides = "top", color = "white")),
            locations = cells_title(groups = "title")) %>%
  tab_options(table.font.size = 12, column_labels.font.size = 12, source_notes.font.size = 12) %>%
  gtsave(here("result", "tables", "Table S2.html"))
```

Convert feature prevalence to boolean values for plotting Venn's diagram
```{r}
core_taxa_venn <- core_taxa %>% 
  rename("REF-DID" = prev_rd, "REF-DIM" = prev_rm, "IM-DID" = prev_id, "IM-DIM" = prev_im) %>%
  mutate_if(is.numeric, ~if_else(.x >= 0.8, 1, 0))
```

Make Venn diagram
```{r }
venn(core_taxa_venn[2:5], ellipse = TRUE, zcolor = brewer.pal(n = 12, name = "Paired")[c(1,2,7,8)], 
     ilcs = 1.2, sncs = 1, box = FALSE)
```

```{r, include=FALSE}
# Export the plot
tiff(here("result", "figures", "Figure 2.tiff"), compression = "lzw", units = "in", res = 300, height = 6, width = 6)
venn(core_taxa_venn[2:5], ellipse = TRUE, zcolor = brewer.pal(n = 12, name = "Paired")[c(1,2,7,8)], 
     ilcs = 1.2, sncs = 1, box = FALSE)
dev.off()
```

# Session information
```{r session-info}
sessionInfo()
```
